File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2023/Y23D19.kt
package org.gristle.pdxpuzzles.advent.y2023

import org.gristle.pdxpuzzles.advent.utilities.Day
import org.gristle.pdxpuzzles.utilities.parsing.blankSplit
import org.gristle.pdxpuzzles.utilities.parsing.getInts

class Y23D19(input: String) : Day {
    
    // A single workflow holds several rules in sequence. 
    // 'Category' is an int representation of the various XMAS categories
    // 'Amount' is used to compare whether there are the correct amount of prats.
    // 'Comparison' can be ">", "<", or "". Blank means automatically pass.
    // 'Destination' is the name of the next workflow to send to if the number/range passes.
    data class Rule(val category: Int, val amount: Int, val comparison: String, val destination: String) 
    
    // Workflows accessible by workflow name.
    private val workflows: Map<String, List<Rule>>
    
    // Parts used for Part 1.
    private val parts: List<List<Int>>
    
    // Parsing 
    init {
        val (workStanza, partStanza) = input.blankSplit()
        workflows = workStanza
            .lines().associate { line ->
                val split = line.split('{', ',')
                val name = split.first()
                val rules = split.drop(1)
                    .map { ruleStr -> // (categoryStr, comparison, amountStr, destination) ->
                        if (ruleStr.last() == '}') {
                            Rule(0, 0, "", ruleStr.dropLast(1))
                        } else {
                            val categoryStr = ruleStr[0]
                            val category = when (categoryStr) {
                                'x' -> 0
                                'm' -> 1
                                'a' -> 2
                                's' -> 3
                                else -> throw IllegalArgumentException("??")
                            }
                            val comparison = ruleStr[1].toString()
                            val amount = ruleStr.dropWhile { !it.isDigit() }.takeWhile { it.isDigit() }.toInt()
                            val destination = ruleStr.takeLastWhile { it.isLetter() }
                            Rule(category, amount, comparison, destination)
                        }
                    }.toList()
                name to rules
            }
        parts = partStanza.getInts()
            .chunked(4)
            .toList()
    }
    
    // For each part, run the recursive sort function on it, which sends it through the various workflows until it 
    // ends up being accepted "A" or rejected "R." Sum the XMAS values for each accepted part and sum the total.
    override fun part1() = parts
        .filter { xmas ->
            sort("in", xmas) == "A"
        }.sumOf { it.sum() }

    private fun sort(name: String, part: List<Int>): String {
        
        // get the relevant workflow
        val workflow = workflows.getValue(name)
        for (rule in workflow) {
            with (rule) {
                
                // perform comparison and route accordingly
                val resultOrNull = if (comparison == ">") {
                    if (part[category] > amount) destination else null
                } else if (comparison == "<") {
                    if (part[category] < amount) destination else null
                } else {
                    destination
                }
                
                // if routed to "A" or "R", we are done. Otherwise call sort again with the next workflow.
                if (resultOrNull == "A" || resultOrNull == "R") {
                    return resultOrNull
                } else if (resultOrNull != null) {
                    return sort(resultOrNull, part)
                }
            }
        }
        return ""
    }
    
    // Start with a PartRanges instance, a list of 4 IntRanges from 1..4000, representing each of X, M, A, and S.
    // Run the instance through the first workflow. Create two PartRanges instances, splitting the ranges for each 
    // branching. Then do the same thing to those instances, until the subranges reach "R" or "A". Keep track of the
    // ones that are accepted, and sum the number of possible permutations 
    override fun part2(): Long {
        val accepted = mutableListOf<PartRanges>()
        var remaining: List<Pair<String, PartRanges>> = listOf("in" to PartRanges(List(4) { 1..4000 }))
        while (remaining.isNotEmpty()) {
            val next = remaining.flatMap { (name, partRanges) ->
                workflows.getValue(name).route(partRanges)
            }
            remaining = next.filter { (name, partRanges) ->
                when (name) {
                    "R" -> false
                    "A" -> {
                        accepted.add(partRanges)
                        false
                    }
                    else -> true
                }
            }
        }
        return accepted.sumOf { it.permutations() }
    }

    // Tracks the ranges for each xmas category.
    @JvmInline
    value class PartRanges(val ranges: List<IntRange>) {
        
        // Number of possible permutations given the four ranges.
        fun permutations(): Long = ranges.fold(1L) { acc, range -> acc * (1 + range.last - range.first) }

        // Splits the PartRanges into two, one that meets the rule condition, and one that doesn't.
        fun split(rule: Rule): List<PartRanges> {
            return when (rule.comparison) {
                ">" -> {
                    val breakpoint = rule.amount + 1
                    val pass = breakpoint..ranges[rule.category].last
                    val fail = ranges[rule.category].first until breakpoint
                    makeSplits(rule.category, pass, fail)
                }
                "<" -> {
                    val breakpoint = rule.amount
                    val pass = ranges[rule.category].first until breakpoint
                    val fail = breakpoint..ranges[rule.category].last
                    makeSplits(rule.category, pass, fail)
                }
                else -> throw IllegalArgumentException("Non-comparisons should not be passed to split function.")
            }
        }
        
        private fun makeSplits(category: Int, pass: IntRange, fail: IntRange): List<PartRanges> =
            listOf(
                PartRanges(List(4) { i -> if (i == category) pass else ranges[i] }),
                PartRanges(List(4) { i -> if (i == category) fail else ranges[i] }),
            )
    }

    // for a particular workflow, return sub-PartRanges along with where they should be routed to next
    private fun List<Rule>.route(partRanges: PartRanges): List<Pair<String, PartRanges>> = buildList {
        var remaining: PartRanges? = partRanges
        for (rule in this@route) {
            if (remaining == null) return@buildList
            remaining = if (rule.comparison.isNotEmpty()) {
                val (pass, fail) = remaining.split(rule)
                add(rule.destination to pass)
                fail
            } else { // every rule conditional failed and base case reached
                add(rule.destination to remaining)
                null
            }
        }
    }
}

fun main() = Day.runDay(Y23D19::class)

//    Class creation: 47ms
//    Part 1: 449531 (4ms)
//    Part 2: 122756210763577 (10ms)
//    Total time: 63ms


// Part 1: 19114	Part 2: 167409079868000
@Suppress("unused")
private val sampleInput = listOf(
    """px{a<2006:qkq,m>2090:A,rfg}
pv{a>1716:R,A}
lnx{m>1548:A,A}
rfg{s<537:gd,x>2440:R,A}
qs{s>3448:A,lnx}
qkq{x<1416:A,crn}
crn{x>2662:A,R}
in{s<1351:px,qqz}
qqz{s>2770:qs,m<1801:hdj,R}
gd{a>3333:R,R}
hdj{m>838:A,pv}

{x=787,m=2655,a=1222,s=2876}
{x=1679,m=44,a=2067,s=496}
{x=2036,m=264,a=79,s=2244}
{x=2461,m=1339,a=466,s=291}
{x=2127,m=1623,a=2188,s=1013}
""",
)

File: rust/advent/src/bin/2023/y23d19.rs
use std::ops::RangeInclusive;

use advent::utilities::get_input::get_input;
use itertools::Itertools;
use lazy_regex::regex;
use rustc_hash::FxHashMap;
use utilities::{parsing::get_numbers::ContainsNumbers, structs::stopwatch::{ReportDuration, Stopwatch}};

type Rules = FxHashMap<String, Vec<Rule>>;
type Input = (Rules, Vec<Vec<usize>>);
type Output = usize;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(23, 19).unwrap();
    let input = parse_input(&input);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

#[derive(Debug)]
struct Rule {
    category: usize,
    amount: usize,
    comparison: Comparison,
    destination: String,
}

#[derive(Debug, PartialEq)]
enum Comparison {
    Greater,
    Less,
    Pass,
}

fn parse_input(input: &str) -> Input {
    let (work_stanza, part_stanza) = input.split_once("\n\n").unwrap();

    let rx = regex!(r"(?<name>[A-z]+)\{(?<conditionals>.*),(?<last>[A-z]+)\}");
    let rx_conditionals = 
        regex!(r"(?<category>[A-z]+)(?<comparison><|>)(?<amount>\d+):(?<destination>[A-z]+)");
    
    let workflows: Rules = work_stanza
        .lines()
        .map(|line| {
            let captures = rx.captures(line).unwrap();
            let name = captures.name("name").unwrap().as_str().to_string();
            let mut conditionals: Vec<Rule> = rx_conditionals
                .captures_iter(captures.name("conditionals").unwrap().as_str())
                .map(|caps| {
                    let category = match caps.name("category").unwrap().as_str() {
                        "x" => 0,
                        "m" => 1,
                        "a" => 2,
                        "s" => 3,
                        c => panic!("{c} not a valid category!"),
                    };
                    let amount: usize = caps.name("amount").unwrap().as_str().parse().unwrap();
                    let comparison = match caps.name("comparison").unwrap().as_str() {
                        "<" => Comparison::Less,
                        ">" => Comparison::Greater,
                        c => panic!("{c} not a valid comparison!"),
                    };
                    let destination = caps.name("destination").unwrap().as_str().to_string();
                    Rule { category, amount, comparison, destination }
                })
                .collect();
            let last = Rule { 
                category: 0, 
                amount: 0, 
                comparison: Comparison::Pass, 
                destination: captures.name("last").unwrap().as_str().to_string(), 
            };
            conditionals.push(last);
            (name, conditionals)
        })
        .collect();

    let parts = part_stanza
        .get_numbers::<usize>()
        .chunks(4)
        .into_iter()
        .map(|chunk| chunk.collect())
        .collect();

    (workflows, parts)
}

fn sort(name: &str, part: &Vec<usize>, workflows: &Rules ) -> String {
    let workflow = workflows.get(name).unwrap();
    for rule in workflow {
        let result = match rule.comparison {
            Comparison::Greater => {
                if part[rule.category] > rule.amount {
                    Some(&rule.destination)
                } else {
                    None
                }
            },
            Comparison::Less => {
                if part[rule.category] < rule.amount {
                    Some(&rule.destination)
                } else {
                    None
                }
            },
            Comparison::Pass => Some(&rule.destination),
        }; 

        if let Some(result) = result {
            return if result == "A" || result == "R" {
                result.clone()
            } else {
                sort(result, part, workflows)    
            }
        } 
    }
    unreachable!()
}

fn part1(input: &Input) -> Output {
    let (workflows, parts) = input;
    parts.iter()
        .filter(|&part| &sort("in", part, workflows) == "A")
        .map(|part| part.iter().sum::<usize>())
        .sum()
}

#[derive(Clone, Debug)]
struct PartRanges([RangeInclusive<usize>; 4]);

impl PartRanges {
    fn new() -> Self {
        Self(std::array::from_fn(|_| 1..=4000))
    }
    
    fn permutations(&self) -> usize {
        self.0.iter().fold(1, |acc, range| {
            acc * (1 + range.end() - range.start())
        })
    }

    fn split(&self, rule: &Rule) -> [PartRanges; 2] {
        match rule.comparison {
            Comparison::Greater => {
                let breakpoint = rule.amount + 1;
                let pass = breakpoint..=*self.0[rule.category].end();
                let fail = *self.0[rule.category].start()..=breakpoint - 1;
                self.make_splits(rule.category, pass, fail)
            },
            Comparison::Less => {
                let breakpoint = rule.amount;
                let pass = *self.0[rule.category].start()..=breakpoint - 1;
                let fail = breakpoint..=*self.0[rule.category].end();
                self.make_splits(rule.category, pass, fail)
            },
            Comparison::Pass => panic!("Non-comparisons should not be passed to split function."),
        }
    }
    
    fn make_splits(&self, 
        category: usize, 
        pass: RangeInclusive<usize>, 
        fail: RangeInclusive<usize>,
    ) -> [PartRanges; 2] {
        [
            PartRanges(std::array::from_fn(|i| {
                if i == category { pass.clone() } else { self.0[i].clone() }})
            ),
            PartRanges(std::array::from_fn(|i| {
                if i == category { fail.clone() } else { self.0[i].clone() }})
            ),
        ]
    }
}

fn route(workflow: &[Rule], part_ranges: PartRanges) -> Vec<(String, PartRanges)> {
    let mut routes = Vec::new();
    let mut remaining = Some(part_ranges);
    for rule in workflow {
        if let Some(range) = remaining {
            remaining = if rule.comparison != Comparison::Pass {
                let [pass, fail] = range.split(rule);
                routes.push((rule.destination.clone(), pass));
                Some(fail)
            } else {
                routes.push((rule.destination.clone(), range));
                None
            }
        } else {
            break;
        }
    }
    routes
}

fn part2(input: &Input) -> Output {
    let (workflows, _) = input;
    let mut accepted: Vec<PartRanges> = Vec::new();
    let mut remaining = Vec::new();
    remaining.push(("in".to_string(), PartRanges::new()));
    while !remaining.is_empty() {
        let next: Vec<_> = remaining.iter()
            .flat_map(|(name, part_ranges)| {
                route(workflows.get(name).unwrap(), part_ranges.clone())
            })
            .collect();
        remaining = next.into_iter()
            .filter(|(name, part_ranges)| {
                match name.as_str() {
                    "R" => false,
                    "A" => { 
                        accepted.push(part_ranges.clone());
                        false 
                    },
                    _ => true,
                }
            })
            .collect();
    }
    accepted.iter().map(|v| v.permutations()).sum()
}

#[test]
fn default() {
    let input = get_input(23, 19).unwrap();
    let input = parse_input(&input);
    assert_eq!(449531, part1(&input));
    assert_eq!(122756210763577, part2(&input));
}

// Input parsed (1ms)
// 1. 449531 (35μs)
// 2. 122756210763577 (247μs)
// Total: 1ms